### PR TITLE
feat: calculate and verify sd-hash

### LINKED ISSUE #8: Key binding JWT does not include add required `_sd_hash` field
See https://datatracker.ietf.org/doc/html/draft-ietf-oauth-selective-disclosure-jwt-06#name-key-binding-jwt

### PR DESCRIPTION
**Pull request by TimoGlastra**
Adds functionality to include the `_sd_hash` as well as verify it.

If you use the top level SdJwt and SdJwtVc classes with `toCompact`, and not add an already signed kb-jwt it will automatically calculate the _sd_hash and include this when signing the kb-jwt.

With verification of SdJwt/SdJwtVc you now need to provide the hasher and algorithm, and it will verify the _sd_hash in the KB JWT. You can also manually set `expectedSdHash` on the key binding class if you don't want the SdJwt or SdJwtVc classes to calculate the expected `_sd_hash`.

I think it should be seen as a breaking change, as now creating KB JWTs without _sd_hash is not possible, and thus the output of this library will change.

Fixes #8

### CODE REVIEW

File: packages/core/src/keyBinding/keyBinding.ts [hunk 1]
```diff
@@ -95,7 +102,15 @@ export class KeyBinding<
...
-        this.assertValidForKeyBinding()
+        if (!this.expectedSdHash) {
+            throw new Error(
+                'Expected sd hash is required for verification of key binding JWT'
+            )
+        }
+
+        // TODO: should _sd_hash also be a verification property (true/false)
+        // or should it throw?
+        await this.assertValidForKeyBinding(this.expectedSdHash)
```
Comment (berendsliedrecht) [inline packages/core/src/keyBinding/keyBinding.ts:113]: We can leave it to throw for now, but I think in the future a verification property would be more inline with the rest of the API.

File: packages/core/src/keyBinding/keyBinding.ts [hunk 2]
```diff
@@ -131,14 +146,26 @@ export class KeyBinding<
         >
     }
 
+    public withSdHash(sdHash: string) {
```
Comment (berendsliedrecht) [inline packages/core/src/keyBinding/keyBinding.ts:149]: Not a 100% whether it would be inline, but for this API I'd prefer the name `withSdHashClaim` as the other `withXXX` methods add a property to the class and not the payload.

File: packages/core/src/sdJwt/sdJwt.ts [hunk 3]
```diff
@@ -209,7 +205,7 @@ export class SdJwt<
      *
      */
     public withKeyBinding(
-        keyBinding: Jwt | KeyBinding | string
+        keyBinding: Jwt | KeyBinding<any, any> | string
```
Comment (berendsliedrecht) [inline packages/core/src/sdJwt/sdJwt.ts:208]: Was this change required?

File: packages/core/src/keyBinding/keyBinding.ts (hunk 1 above)
Comment (TimoGlastra) [reply packages/core/src/keyBinding/keyBinding.ts:113]: Ah yeah i was really doubting as the other checks (like whether `iat` is present) is not part of the boolean result values but will throw an error directly. Should we also move these to the validation object?

I think it's nice if it will always return a result object rather than sometimes throwing

File: packages/core/src/keyBinding/keyBinding.ts (hunk 2 above)
Comment (TimoGlastra) [reply packages/core/src/keyBinding/keyBinding.ts:149]: Makes sense 👍

File: packages/core/src/sdJwt/sdJwt.ts (hunk 3 above)
Comment (TimoGlastra) [reply packages/core/src/sdJwt/sdJwt.ts:208]: Yes, otherwise it will complain that `KeyBinding<Payload, Header>` is not compatible with `Keybinding<Record<string, unknown>, Record<string, unknown>>` (as that will be the default if you don't pass any)

**Comment by TimoGlastra** [comment]
I updated the claim name, the others are OK to fix later right?
